### ISSUE TITLE
Kubelet should enforce that static pods do not reference other API objects

### ISSUE THREAD

**Issue by MikeSpreitzer** [kubernetes/kubernetes#103587]
#### What happened:
I put into `/etc/kubernetes/manifests/` one defining a static pod with a `serviceAccount`, because I was wondering if this is supported (the documentation does not say, see https://github.com/kubernetes/website/issues/28611#issuecomment-876537379).  The kubelet created the pod but without giving it the relevant mount.  The kubelet did _not_ create the corresponding mirror pod.

Here is the manifest I created:
```
apiVersion: v1
kind: Pod
metadata:
  name: test5
spec:
  containers:
  - name: bb
    image: busybox
    command:
    - sleep
    - "864001"
  serviceAccountName: default
```
Here are some relevant log entries from the kubelet:
```
Jul 08 15:55:05 init1 kubelet[4229]: I0708 15:55:05.335792    4229 kubelet.go:1681] "Creating a mirror pod for static pod" pod="default/test5-init1"

Jul 08 15:55:05 init1 kubelet[4229]: E0708 15:55:05.343771    4229 kubelet.go:1683] "Failed creating a mirror pod for" err="pods \"test5-init1\" is forbidden: a mirror pod may not reference service accounts" pod="default/test5-init1"
```

#### What you expected to happen:
I expected consistent treatment --- real pod and mirror pod both exist or both do not exist.

#### How to reproduce it (as minimally and precisely as possible):
I think the above is pretty minimal.

#### Environment:
- Kubernetes version (use `kubectl version`):
  ```
  Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:18:45Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
   Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
  ```
- Cloud provider or hardware configuration: VirtualBox x86_64 VMs with 2 virtual CPUs and 2 GB of memory
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04.1 LTS
- Kernel (e.g. `uname -a`): `Linux init1 5.4.0-65-generic #73-Ubuntu SMP Mon Jan 18 17:25:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: kubeadm v1.21.2
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-node-bugs

**Comment by liggitt** [comment]
It is expected that static pod will not run... no references to API objects are permitted in static pods (service accounts, PVC objects, secrets, configmaps, etc). 

> The kubelet created the pod but without giving it the relevant mount. The kubelet did not create the corresponding mirror pod.

I'm not sure I understand what you mean by this

**Comment by MikeSpreitzer** [comment]
Ooh, the prohibition is even broader.  @liggitt please have a look at https://github.com/kubernetes/website/pull/28623 and there should probably also be updates to PodSpec and/or types inside it.

As for what I meant: the kubelet implemented the pod:
```
root@init1:~# docker ps
CONTAINER ID   IMAGE                    COMMAND                  CREATED             STATUS             PORTS     NAMES
c87ee60d1f92   busybox                  "sleep 864001"           About an hour ago   Up About an hour             k8s_bb_test5-init1_default_595a2ca0aaa264ec6f10330a8d5162a8_0
6c885e29da59   k8s.gcr.io/pause:3.4.1   "/pause"                 About an hour ago   Up About an hour             k8s_POD_test5-init1_default_595a2ca0aaa264ec6f10330a8d5162a8_0
...
```

but the mirror pod object did not get created in the apiserver.

**Comment by endocrimes** [comment]
Definitely seems like a bug that we actually run that - given that we can never report it to the API.

/triage accepted

**Comment by endocrimes** [comment]
/remove-lifecycle stale
/lifecycle-frozen

(valid bug that is unlikely to be fixed by accident)

**Comment by enj** [comment]
/retitle Kubelet should enforce that static pods do not reference other API objects

xref: https://github.com/kubernetes/kubernetes/pull/105368#issuecomment-1215099405

**Comment by enj** [comment]
/lifecycle frozen
/sig storage

Adding SIG storage since volume mounts outside of host path and empty dir do not make sense for static pods.

**Comment by enj** [comment]
xref: https://github.com/kubernetes/website/pull/28623 https://github.com/kubernetes/website/issues/28611 #103167

**Comment by liggitt** [comment]
the HEAD commit at https://github.com/liggitt/kubernetes/commits/kubelet-static-api-reference is a sketch of what this could look like, catching API references in file and url sourced pods and preventing them from ever being handled by the kubelet

**Comment by smarterclayton** [comment]
/assign

Agreed this should be blocked at creation time

**Comment by smarterclayton** [comment]
Is this enforced at API today?  Is there any backward compatibility implication to tightening Kubelet validation?

**Comment by liggitt** [comment]
> Is this enforced at API today?

Yes, NodeRestriction admission blocks the serviceaccount/secret/configmap/pvc references that would inform node authorizer:

https://github.com/kubernetes/kubernetes/blob/0270fc75d0f59e1f5a31a810737ac35ac3eeb8a3/plugin/pkg/admission/noderestriction/admission.go#L247-L265

Blocking pod creation there means the node is silently running static pods that *probably* aren't working (API references that require authz fail unless another pod on the same node happens to reference the same API object), and *definitely* are confusing the scheduler because the resources they consume aren't represented in the API.

> Is there any backward compatibility implication to tightening Kubelet validation?

A kubelet could conceivably be running static pods whose referenced resources *happened* to also be referenced by a non-static pod on the same node, but that doesn't seem like it would result in consistently successful pods.

**Comment by liggitt** [comment]
xref https://github.com/kubernetes/kubernetes/pull/113374 which is adding another API reference field to protect against in static / mirror pods

**Comment by haircommander** [comment]
/remove-sig node
/sig architecture
/triage needs-information
/priority backlog

We need some advisement on how to proceed with this. changes like this create backwards compatibility issues. It's possible we shouldn't address this in tree, and maybe recommend a daemonset or something that edits the manifests. Either way, we need some guidance and probably can't decide this alone

**Comment by yujuhong** [comment]
> We need some advisement on how to proceed with this. changes like this create backwards compatibility issues. It's possible we shouldn't address this in tree, and maybe recommend a daemonset or something that edits the manifests. Either way, we need some guidance and probably can't decide this alone

This issue seems like a bug and the created static pods may not be working 100% anyway. 
If we want to be extra careful, we could try one of the following routes:
 - Add a warning but not fail for one release to raise awareness; fail the validation in the next release
 - Fail the validation, but add a temporary flag to restore the old behavior to tie user over for a release or two (of course then we'll have to remove the flag later). 
 
WDYT? @haircommander

**Comment by liggitt** [comment]
> Fail the validation, but add a temporary flag to restore the old behavior to tie user over for a release or two (of course then we'll have to remove the flag later).

That's what I would expect.

**Comment by AnishShah** [comment]
sig-node CI meeting:

/triage accepted

**Comment by sreeram-venkitesh** [comment]
> make the tightened validation able to be opted out of with a feature gate marked as deprecated for a release or two

We need a KEP in this case right?

**Comment by liggitt** [comment]
> > make the tightened validation able to be opted out of with a feature gate marked as deprecated for a release or two
> 
> We need a KEP in this case right?

I think this falls into the "bugfix with gate as an escape hatch" category, I don't think this needs a KEP
